pkg/kademlia/dht.go

```diff
@@ -5,37 +5,178 @@ package kademlia
 
 import (
 	"context"
+	"crypto/rand"
+	"fmt"
+	"strings"
 
-	"storj.io/storj/protos/overlay"
+	bkad "github.com/coyle/kademlia"
+	"github.com/zeebo/errs"
+
+	proto "storj.io/storj/protos/overlay"
 )
 
+// NodeErr is the class for all errros petaining to node operations
+var NodeErr = errs.Class("node error")
+
+//TODO: shouldn't default to TCP but not sure what to do yet
+var defaultTransport = proto.NodeTransport_TCP
+
 // Kademlia is an implementation of kademlia adhering to the DHT interface.
 type Kademlia struct {
+	rt             RoutingTable
+	bootstrapNodes []proto.Node
+	ip             string
+	port           string
+	stun           bool
+	dht            *bkad.DHT
+}
+
+// NewKademlia returns a newly configured Kademlia instance
+func NewKademlia(bootstrapNodes []proto.Node, ip string, port string, stun bool) Kademlia {
+	bb := convertProtoNodes(bootstrapNodes)
+	bdht, _ := bkad.NewDHT(&bkad.MemoryStore{}, &bkad.Options{
+		ID:             []byte(newID()),
+		IP:             ip,
+		Port:           port,
+		BootstrapNodes: bb,
+	})
+
+	rt := RouteTable{
+		ht:  bdht.HT,
+		dht: bdht,
+	}
+
+	return Kademlia{
+		rt:             rt,
+		bootstrapNodes: bootstrapNodes,
+		ip:             ip,
+		port:           port,
+		stun:           stun,
+		dht:            bdht,
+	}
 }
 
-// GetNodes returns all nodes from a starting node up to a maximum limit
-func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]overlay.Node, error) {
-	return []overlay.Node{}, nil
+// GetNodes returns all nodes from a starting node up to a maximum limit stored in the local routing table
+func (k Kademlia) GetNodes(ctx context.Context, start string, limit int) ([]proto.Node, error) {
+	nn, err := k.dht.FindNodes(ctx, start, limit)
+	if err != nil {
+		return []proto.Node{}, err
+	}
+	return convertNetworkNodes(nn), nil
 }
 
 // GetRoutingTable provides the routing table for the Kademlia DHT
 func (k Kademlia) GetRoutingTable(ctx context.Context) (RoutingTable, error) {
-	return RouteTable{}, nil
+	return RouteTable{
+		ht:  k.dht.HT,
+		dht: k.dht,
+	}, nil
 }
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k Kademlia) Bootstrap(ctx context.Context) error {
-	return nil
+	if err := k.dht.CreateSocket(); err != nil {
+		return err
+	}
+
+	go k.dht.Listen()
```

Comment:
i feel like this is one of those things where the creator of the Kademlia struct should expect to kick off some kind of "ListenAndServe" on their own, instead of it happen automatically as part of another function call?

Reply:
Are you saying provide a function ListenAndServe that can be called by the users of the kademlia package ?

Reply:
yeah